src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs

```diff
@@ -1,156 +1,179 @@
+                        newQueryString = MergeQueryStringsWithoutDuplicateValues(downstreamRequest.Query, newQueryString);
+                        downstreamRequest.Query = "?" + newQueryString;                    }
+                }
+                else
+                {
+                    RemoveQueryStringParametersThatHaveBeenUsedInTemplate(downstreamRequest, templatePlaceholderNameAndValues);
+
+                    downstreamRequest.AbsolutePath = dsPath.Value;
+                }
+            }
+
+            Logger.LogDebug($"Downstream url is {downstreamRequest}");
+
+            await _next.Invoke(httpContext);
+        }
+
+        private static string MergeQueryStringsWithoutDuplicateValues(string queryString, string newQueryString)
+        {
+            var queries = HttpUtility.ParseQueryString(queryString);
+            var newQueries = HttpUtility.ParseQueryString(newQueryString);
+
+            var dict = new Dictionary<string, string>();
+            foreach (var key in newQueries.AllKeys)
+            {
+                if (!string.IsNullOrEmpty(key))
+                {
+                    dict.Add(key, newQueries[key]);
+                }
+            }
+
+            foreach (var key in queries.AllKeys)
+            {
+                if (string.IsNullOrEmpty(key) || dict.ContainsValue(queries[key]))
+                {
+                    continue;
+                }
+
+                dict.Add(key, queries[key]);
+            }
+
+            return string.Join("&", dict.Select(kvp => string.Format("{0}={1}", kvp.Key, kvp.Value)));
+        }
+
+        private static void RemoveQueryStringParametersThatHaveBeenUsedInTemplate(DownstreamRequest downstreamRequest, List<PlaceholderNameAndValue> templatePlaceholderNameAndValues)
+        {
+            foreach (var nAndV in templatePlaceholderNameAndValues)
+            {
+                var name = nAndV.Name.Replace("{", string.Empty).Replace("}", string.Empty);
+
+                if (downstreamRequest.Query.Contains(name) &&
+                    downstreamRequest.Query.Contains(nAndV.Value))
+                {
+                    var questionMarkOrAmpersand = downstreamRequest.Query.IndexOf(name, StringComparison.Ordinal);
+                    downstreamRequest.Query = downstreamRequest.Query.Remove(questionMarkOrAmpersand - 1, 1);
+
+                    var rgx = new Regex($@"\b{name}={nAndV.Value}\b");
+                    downstreamRequest.Query = rgx.Replace(downstreamRequest.Query, string.Empty);
+
+                    if (!string.IsNullOrEmpty(downstreamRequest.Query))
+                    {
+                        downstreamRequest.Query = '?' + downstreamRequest.Query.Substring(1);
+                    }
+                }
+            }
+        }
+
+        private static string GetPath(DownstreamPath dsPath)
+        {
+            return dsPath.Value.Substring(0, dsPath.Value.IndexOf('?', StringComparison.Ordinal));
```

Comment:
'?' could a private const of class

Reply:
Fixed in https://github.com/ThreeMammals/Ocelot/pull/1182/commits/275d034250762557f50e8461d27b3446893b54a1